db/dbScript.sql

```diff
@@ -0,0 +1,393 @@
+BEGIN
+    IF NOT EXISTS (SELECT 1 FROM pg_type WHERE typname = 'provider_type') THEN
+        CREATE TYPE provider_type AS ENUM ('local', 'google');
+    END IF;
+	IF NOT EXISTS (SELECT 1 FROM pg_type WHERE typname = 'account_type') THEN
+        CREATE TYPE account_type AS ENUM ('user', 'gym', 'admin');
+    END IF;
+	IF NOT EXISTS (SELECT 1 FROM pg_type WHERE typname = 'own_status') THEN
+        CREATE TYPE own_status AS ENUM ('pending', 'approved', 'rejected');
+    END IF;
+	IF NOT EXISTS (SELECT 1 FROM pg_type WHERE typname = 'rec_type') THEN
+        CREATE TYPE rec_type AS ENUM ('main', 'alternative');
+    END IF;
+	IF NOT EXISTS (SELECT 1 FROM pg_type WHERE typname = 'not_type') THEN
+        CREATE TYPE not_type AS ENUM ('message', 'alert', 'reminder');
+    END IF;
+END $$;
+
+--Since we will have a lot of gyms stored (potentially) it does not make sense to store working hours for each of them as a text
+--Especially since there are significantly smaller potential patterns [(opening hours, closing hours) pairs] in comparison to the potential number of gyms stored
+CREATE TABLE if not exists Working_hours
+(
+  id uuid NOT null DEFAULT uuid_generate_v4(),--uuid is analogous to id, but uuid is pseudo-random and the length and format of uuids are the same
+  open_from TIME NOT NULL,
+  open_until TIME NOT NULL,
+  PRIMARY KEY (id),
+  UNIQUE (open_from),
+  UNIQUE (open_until),
+  check (open_until > open_from)
+);
+
+CREATE TABLE if not exists Currency
+(
+  id uuid NOT null DEFAULT uuid_generate_v4(),
+  name VARCHAR(10) NOT NULL,
+  code CHAR(3) NOT NULL,
+  PRIMARY KEY (id),
+  UNIQUE (name),
+  UNIQUE (code)
+);
+
+CREATE TABLE if not exists Account
+(
+  id uuid NOT null DEFAULT uuid_generate_v4(),
+  username VARCHAR(40) NOT null check (length(username) > 5),
+  email VARCHAR(60) not NULL,
+  first_name VARCHAR(60) NOT NULL,
+  last_name VARCHAR(60) NOT NULL,
+  created_at TIMESTAMP NOT null check (created_at <= NOW()),
+  provider provider_type NOT NULL, --we want to differentiate between google accounts and local accounts
+  password_hash CHAR(60) NOT NULL, --we are storing hash of the password that we got from bycrypt
+  type account_type NOT NULL, --we want to differentate between user, admin and gym accounts
+  created_by uuid, --admin accounts can be created only by another admin accounts. in this case we want to register who created the admin account
+  PRIMARY KEY (id),
+  FOREIGN KEY (created_by) REFERENCES Account(id),
+  UNIQUE (username),
+  UNIQUE (email)
+);
+create index if not exists idx_account_username on Account(username);
+create index if not exists idx_account_email on Account(email);
+create index if not exists idx_account_type on Account("type");
+
+create table if not exists Country
+(
+	id uuid not null default uuid_generate_v4(),
+	name varchar(56) not null,
+	primary key(id),
+	UNIQUE (name)
+);
+create index if not exists idx_country_name on Country("name");
+
+create table if not exists City
```

Comment:
Why do we need a separate **city** table? We could store the city name as well as country name directly in the **gym** table for search queries. What is the use case where city coordinates are necessary?

Reply:
The point was to divide the rectangular approximation of the city area provided by Google API into multiple round sectors (because Google Places Nearby API requires you to specify a radius for which it will retrieve data) in order to get as many gyms in the city as possible (using Google Places API in way that we request sth like "gyms Zagreb Croatia" will filter out some less popular gyms because of the way the API works. I wanted to avoid it this way)

This may be an overkill though, what do you think?

Reply:
Okay, we will approach it slightly differently, but it is not an overkill. We probably need the coordinates just for the case where we have two cities/towns with the same name in one country. Then we would have to compare the coordinates as well.